### ISSUE TITLE
Question on using a sequence length > the max length I can hold in a batch due to memory usage for training

### ISSUE THREAD

**Issue by luke-mcdermott-mi** [alxndrTL/mamba.py#40]
For running extremely large sequence lengths, I can break up the sequence among batches and use the inference step to save the hidden states between passes. However, step does not work during training as there seems to be a problem with overwriting the variables before loss.backward() can be run.

How might you modify the forward pass to allow running the same parallel scan as used during training, but connect previous hidden states from other batches?

**Comment by alxndrTL** [comment]
That's an interesting problem

I guess you would have to make N sequential parallel scans, each one starting with a hidden state equals to the last hidden state of the previous pscan.

Just to make sure we understand each other :
![IMG_0220](https://github.com/alxndrTL/mamba.py/assets/22843596/82586122-abc0-47e4-9263-34b835640924)

Concerning starting a pscan with a hidden state that is not zero (the code supposed that `H[0] = 0`), I recently answered an [issue about that](https://github.com/alxndrTL/mamba.py/issues/38).

Hope that helps

**Comment by alxndrTL** [comment]
Hello again, just wanted to inform you that I have started working on your question, you can follow the progress on the [mem-update branch](https://github.com/alxndrTL/mamba.py/tree/mem-update). Splitting in chunks (like you asked) is very useful to reduce the memory usage, and it has been a long time I wanted to do a "memory update" for mamba.py so :)

**Comment by MilkClouds** [comment]
@alxndrTL , the mem-update branch you implement seems awesome and will be very helpful also to me. Is there any specific reason why it's not merged into `main`?
